proxy/src/compute.rs

```diff
@@ -110,7 +112,38 @@ pub struct PostgresConnection {
 
 impl NodeInfo {
     /// Connect to a corresponding compute node.
-    pub async fn connect(&self) -> Result<(PostgresConnection, CancelClosure), ConnectionError> {
+    pub async fn connect(
+        mut self,
+        params: &StartupMessageParams,
+    ) -> Result<(PostgresConnection, CancelClosure), ConnectionError> {
+        if let Some(options) = params.options_raw() {
+            // We must drop all proxy-specific parameters.
+            #[allow(unstable_name_collisions)]
+            let options: String = options
+                .filter(|opt| !opt.starts_with("project="))
+                .intersperse(" ") // TODO: use impl from std once it's stabilized
+                .collect();
+
+            self.config.options(&options);
+        }
+
+        if let Some(app_name) = params.get("application_name") {
+            self.config.application_name(app_name);
+        }
+
+        if let Some(replication) = params.get("replication") {
+            use tokio_postgres::config::ReplicationMode;
+            match replication {
+                "true" | "on" | "yes" | "1" => {
+                    self.config.replication_mode(ReplicationMode::Physical);
+                }
+                "database" => {
+                    self.config.replication_mode(ReplicationMode::Logical);
+                }
+                _other => {}
+            }
+        }
+
         let (socket_addr, mut stream) = self
```

Comment:
In principle, the startup packet could contain arbitrary GUCs directly as parameters. Per the PostgreSQL docs, on the [StartupMessage format](https://www.postgresql.org/docs/current/protocol-message-formats.html):

> In addition to the above, other parameters may be listed. Parameter names beginning with `_pq_.` are reserved for use as protocol extensions, while others are treated as run-time parameters to be set at backend start time. Such settings will be applied during backend start (after parsing the command-line arguments if any) and will act as session defaults.

Not sure which drivers support that in practice. I see that the tokio-postgres doesn't allow passing arbitrary params to the StartupMessage that it constructs, so passing through others than the ones listed here would need some patching of tokio-postgres. Let's at least add a comment on that.

Reply:
Thank you, I'm aware of that! As you've mentioned, this requires more changes to the underlying driver. I enabled arguably the most important ones, so let's take one step at I time. Meanwhile, I've added a TODO.